bin/varnishd/cache/cache_hash.c

```diff
@@ -369,7 +369,9 @@ hsh_rush_match(struct req *req)
 	CHECK_OBJ_NOTNULL(oc, OBJCORE_MAGIC);
 
 	AZ(oc->flags & OC_F_BUSY);
-	if (oc->flags)
+	if (oc->flags & (OC_F_WITHDRAWN|OC_F_HFM|OC_F_HFP|OC_F_CANCEL|
```

Comment:
Why are HFM and HFP being excluded here? Aren't they supposed to explicitly do the rushing?

Reply:
At this point the request has already been rushed, and it's failing to match an objcore with either HFP or HFM, so it will proceed with a regular lookup. In all likelihood it will proceed in either `vcl_pass` or `vcl_miss` accordingly, unless another compatible object landed in the cache between rush time and lookup time (when we grab the objhead lock).

Reply:
I see it now how that comes together, at least for HFP objects. The rushing cascade will continue in `HSH_Lookup()` when the HFP flag is found and `hsh_deref_objhead_unlock()` is called at https://github.com/Dridi/varnish-cache/blob/objcore_rush/bin/varnishd/cache/cache_hash.c#L556. `hsh_deref_objhead_unlock()` does a wake of `rush_exponent` tasks when finding a HFP object.

Though I worry that there is no similar effect for HFM with this patch set. For those, only `rush_exponent` tasks will be woken during `HSH_Unbusy()`. But these will not rush again during their `HSH_Lookup()` (https://github.com/Dridi/varnish-cache/blob/objcore_rush/bin/varnishd/cache/cache_hash.c#L581), pausing the cascade. It will wake again when one of those does `HSH_Unbusy()`.

Reply:
HFM will cascade, but not as soon as HFP. Unlike HFP, HFM may break the exponential cascade.

Since we treat it as a miss, and retain the possibility of inserting a cacheable object on the next fetch, we will either have a wholesale rush (cacheable) or an exponential rush at "unbusy" time (if all `rush_exponent` fetch tasks trigger `rush_exponent` waiting list rushes of their own).

Reply:
Checking this again, I guess this behaviour of HFM is the same as before this PR, and with that knowledge I don't have any objection to this PR.

Though I was caught by surprise wrt to the delay inherent to the HFM rushing. My understanding of this was that HFM would behave the same as HFP. The rushing is only used to space out the task scheduling internal to Varnish, and not a method of being gentle with the backend. I think it's worth revisiting this logic regardless of this PR, and seeing if we have the intended behaviour here.